### ISSUE TITLE
Host the database on a server to access it from more devices

### ISSUE THREAD

**Issue by Pietrob5** [Pietrob5/Password-Manager#13]
I'd like to give users the ability to upload their database to a server. Since private data is encrypted, the database should be safe to store online (Correct me if I'm wrong).
I believe the easiest implementation is to download the database every time a user opens the application, allowing access for reading or writing, and then upload it back to the server. This is not the most efficient way (how big can a string-based database be?), but it is simpler than sending instructions for every user input.
I haven't yet decided how to authenticate the users (to ensure that a user can only download their own database and not others).
For any questions or suggestions, contact me regarding this issue.
Security must be the priority.

**Comment by Luquor** [comment]
You could hash the password with python before storing it to the database. It is a more efficient and 'classical' way of doing things.
Here is a link to python's documentation about hashlib: https://docs.python.org/3/library/hashlib.html

EDIT: if there are multiple users, you can assure to only access the user's passwords by adding a table that maps the id of the password with the id of a user.

**Comment by Pietrob5** [comment]
> You could hash the password with python before storing it to the database. 

I'm already doing it. I don't understand

> EDIT: if there are multiple users, you can assure to only access the user's passwords by adding a table that maps the id of the password with the id of a user.

You mean to create a table that link userID to the name of his database? For example an entry like (User1, abc.db)?

**Comment by Luquor** [comment]
My bad I didn't understand the problem the first time, I thought you wanted to store the password in a database, not store the users' database!

> You mean to create a table that link userID to the name of his database? For example an entry like (User1, abc.db)?

Yes you could do that it should be doing the job.

> I'm already doing it. I don't understand

So depending on what kind of database you are using for the project (it seems to be SQLITE), you could send the database **only** when there is a change (the user add, update, delete a password from the manager). In that case, you can limit the amount of data sent; it means that the user will always have a local version of the database.

**Comment by Pietrob5** [comment]
> you could send the database only when there is a change 

I was thinking about that when a user opens the application, it automatically downloads its database on the local machine. Then the user can search, add, delete a password and when he closes the application the database (updated) is sent back to the server. I think this is the easyest way to do it (but not the most efficient).

**Comment by Luquor** [comment]
And is it not possible to, instead of uploading/downloading the database, to make the requests directly to that database (on the server you want to upload database)?
You cannot authenticate the user when he launches the application so he only has access to his passwords. That exclude the downloading/uploading process.

**Comment by Pietrob5** [comment]
@Luquor Yes I could but it's more complex to implement, isn't it?

**Comment by i3hz** [comment]
Welp im a newbie in python so I could be wrong but couldn't you just host a server using mysql and connect to a database there

**Comment by Pietrob5** [comment]
> Welp im a newbie in python so I could be wrong but couldn't you just host a server using mysql and connect to a database there

Explain your plan please

**Comment by Luquor** [comment]
> Yes I could but it's more complex to implement, isn't it?

As @i3hz said, you can host a mysql/mariadb/postgresql (you have to find the database solution that fits your needs) either on a platform or with docker. You can still use SQLITE btw, it just might not be the more appropriate solution.
Like for SQLITE, you need to use a driver to be able to connect python with your database.
Having a good SQL architecture is the hardest part here. I, personally, will do this:

- A table for the users will have:
  - ID
  - userID (not mandatory)
  - e-mail (if it's useful in your case)
  - master password to unlock the manager 
- A table for the passwords will have:
  - ID
  - userID (ID of the user that created the password)
  - hashed password
  - And maybe a date when created and last updated.
  
The requests and the logic of your application can still be the same as before (modulo the type fo database you use).

**Comment by Pietrob5** [comment]
@Luquor you are more than welcome to clone the repository and have a look at my database. If you want you can also write some base code to host it online (also without the authentication part, just to test it)

**Comment by i3hz** [comment]
Yeah pretty much what @Luquor  said , i did a small ml project where I had to store attendance on a webpage , it was pretty basic and locally hosted but you could probably do something along those lines to create a database online

**Comment by Pietrob5** [comment]
Do you guys want to help me?

**Comment by Luquor** [comment]
Yes I will help you, no problem!
I will clone the repo and make some modifications to implement what we discussed earlier; I will add a markdown file that explain to you the modifications, how to setup your database etc.
By the time that I make the changes (one or two days), I linked to you a repo that I made few years ago when I was a student in CS (just like you know). Even if it is in PHP, the code is quite simple to apprehend; the part that is the most useful is the file in the database, and in the classes directory the file called authentification. Github repo: https://github.com/Luquor/NetVOD
